packages/block-library/src/gallery/test/index.native.js

```diff
@@ -27,102 +69,578 @@ afterAll( () => {
 	} );
 } );
 
-const GALLERY_WITH_ONE_IMAGE = `<!-- wp:gallery {"linkTo":"none"} -->
-<figure class="wp-block-gallery has-nested-images columns-default is-cropped"><!-- wp:image {"id":1} -->
-<figure class="wp-block-image"><img src="https://cldup.com/cXyG__fTLN.jpg" alt="" class="wp-image-1"/></figure>
-<!-- /wp:image --></figure>
-<!-- /wp:gallery -->`;
+describe( 'Gallery block', () => {
+	it( 'inserts block', async () => {
+		const { getByA11yLabel } = await addGalleryBlock();
 
-const addGalleryBlock = async () => {
-	const screen = await initializeEditor();
-	const { getByA11yLabel, getByTestId, getByText } = screen;
+		expect( getByA11yLabel( /Gallery Block\. Row 1/ ) ).toBeVisible();
+		expect( getEditorHtml() ).toMatchSnapshot();
+	} );
 
-	fireEvent.press( getByA11yLabel( 'Add block' ) );
+	it( 'selects a gallery item', async () => {
+		const { galleryBlock } = await initializeWithGalleryBlock( {
+			numberOfItems: 1,
+			media,
+			selected: false,
+		} );
+
+		const galleryItem = getGalleryItem( galleryBlock, 1 );
+		fireEvent.press( galleryItem );
 
-	const blockList = getByTestId( 'InserterUI-Blocks' );
-	// onScroll event used to force the FlatList to render all items
-	fireEvent.scroll( blockList, {
-		nativeEvent: {
-			contentOffset: { y: 0, x: 0 },
-			contentSize: { width: 100, height: 100 },
-			layoutMeasurement: { width: 100, height: 100 },
-		},
+		expect( galleryItem ).toBeVisible();
 	} );
 
-	fireEvent.press( await waitFor( () => getByText( 'Gallery' ) ) );
+	it( 'shows appender button when gallery has images', async () => {
+		const { galleryBlock, getByText } = await initializeWithGalleryBlock( {
+			numberOfItems: 1,
+			media,
+		} );
 
-	return screen;
-};
+		const appenderButton = within( galleryBlock ).getByA11yLabel(
+			/Gallery block\. Empty/
```

Comment:
I was wondering if we could use something more specific to target the appender button (+) instead of the whole placeholder. From the recent regression we had, I thought about what would happen if [this piece of code was removed](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/gallery/edit.js#L439-L446). I was expecting a failing test but all passed. It's unlikely that some mobile-specific code would be removed without asking but who knows 😄. In this case, it passed because the placeholder was visible and working, but the content of it was wrong:

<kbd><img width="150" src="https://user-images.githubusercontent.com/4885740/154301181-f05b9782-8a60-4257-9687-a6891f590394.png" /></kbd>

Maybe in `MediaPlaceholder` we could add a specific testID for the `renderContent` wrapper depending on if `isAppender` is true or not or something like that.

What do you think? Is it too much? My thinking is we should diferentiate it 😅

Reply:
Good point, I thought that I had tested that flow 😅. Yeah, I agree that we should rely on something more specific that would give us the confidence that, if the appender button is disabled, the test would fail 👍 . Addin a `testID` would work so I'm most likely leaning towards that option, thanks for raising this 🙇 .

Reply:
> I thought that I had tested that flow 😅

If you apply the following patch, following the changes from https://github.com/WordPress/gutenberg/pull/38238, the test actually fails.

```diff
diff --git a/packages/block-library/src/gallery/edit.js b/packages/block-library/src/gallery/edit.js
index 99bd651c69d9da78141f28ae799b9e905641da1e..78aaf5e0170fd9afdb455a0863b5e1a7ba53c86c 100644
--- a/packages/block-library/src/gallery/edit.js
+++ b/packages/block-library/src/gallery/edit.js
@@ -437,13 +437,9 @@ function GalleryEdit( props ) {
 			value: {},
 		},
 		native: {
-			addToGallery: hasImageIds,
-			isAppender: hasImages,
-			disableMediaButtons:
-				( hasImages && ! isSelected ) || imagesUploading,
-			value: hasImageIds ? images : {},
-			autoOpenMediaUpload:
-				! hasImages && isSelected && wasBlockJustInserted,
+			addToGallery: false,
+			disableMediaButtons: imagesUploading,
+			value: {},
 		},
 	} );
 	const mediaPlaceholder = (
@@ -551,11 +547,7 @@ function GalleryEdit( props ) {
 			<Gallery
 				{ ...props }
 				images={ images }
-				mediaPlaceholder={
-					! hasImages || Platform.isNative
-						? mediaPlaceholder
-						: undefined
-				}
+				mediaPlaceholder={ ! hasImages ? mediaPlaceholder : undefined }
 				blockProps={ blockProps }
 				insertBlocksAfter={ insertBlocksAfter }
 			/>

```

However, it's true that if `isAppender` prop is set to `false`, the test gives a false positive, although it's not strictly false as it's displaying a button to append a media item 😅 .

<img width="422" alt="Screenshot 2022-02-17 at 18 47 35" src="https://user-images.githubusercontent.com/14905380/154540513-11268192-910a-4c5a-a788-60b81f3be385.png">

In any case, I like the idea of adding a `testID` to be more accurate on the test assertion. I'll push a commit for this.

Reply:
I applied the mentioned suggestion in https://github.com/WordPress/gutenberg/pull/38571/commits/c1e37ad1a6387a8a3e4c39f134fc63c4988e08c4.

I'd like to note that I had to add a `View` component just for the purpose of adding the test ID. To be honest, it's not a good practice modifying the business logic for testing purposes but I understand that in this case, due to the way the appender icon is structured, it was necessary. I'd appreciate it @geriux you could take a look and provide your opinion, thanks 🙇 .

Reply:
> If you apply the following patch, following the changes from https://github.com/WordPress/gutenberg/pull/38238, the test actually fails.

Right, what I tried was removing the whole `native` part, which I thought was most likely to happen 😄 

> I'd like to note that I had to add a View component just for the purpose of adding the test ID. To be honest, it's not a good practice modifying the business logic for testing purposes but I understand that in this case, due to the way the appender icon is structured, it was necessary. 

Yeah, I agree that we shouldn't modify the business logic, but as you said for this case I believe is needed to not rely on asserting for a label like `/Gallery block\. Empty/` to assert for the placeholder (+) button.

Changes look good to me! Thanks for updating it.